contracts/atomic-swap/src/msg.rs

```diff
@@ -0,0 +1,84 @@
+use cosmwasm_std::{Coin, HumanAddr};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, JsonSchema)]
+pub struct InitMsg {}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum HandleMsg {
+    Create(CreateMsg),
+    /// Release sends all tokens to the recipient.
+    Release {
+        id: String,
+        /// This is the preimage, must be exactly 32 bytes in hex (64 chars)
+        /// to release: sha256(from_hex(preimage)) == from_hex(hash)
+        preimage: String,
+    },
+    /// Refund returns all remaining tokens to the original sender,
+    Refund {
+        id: String,
+    },
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct CreateMsg {
+    /// id is a human-readable name for the swap to use later.
+    /// 3-20 bytes of utf-8 text
+    pub id: String,
+    /// This is hex-encoded sha-256 hash of the preimage (must be 32*2 = 64 chars)
+    pub hash: String,
+    /// If approved, funds go to the recipient
+    pub recipient: HumanAddr,
+    /// You can set a last time or block height the contract is valid at.
+    /// If *either* is non-zero and below current state, the contract is considered expired,
+    /// and will be returned to the original funder.
+    pub end_height: u64,
+    pub end_time: u64,
+}
+
+pub fn is_valid_name(name: &str) -> bool {
+    let bytes = name.as_bytes();
+    if bytes.len() < 3 || bytes.len() > 20 {
+        return false;
+    }
+    true
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum QueryMsg {
+    /// Show all open swaps. Return type is ListResponse.
```

Comment:
This obviously will need pagination. Look at https://github.com/CosmWasm/cosmwasm-plus/pull/64 where I add some iterators to cw20-base. Keep point is the pagination.

Pagination input (2 fields on query): https://github.com/CosmWasm/cosmwasm-plus/blob/master/contracts/cw20-base/src/msg.rs#L149-L150

Logic to apply the start/limit: https://github.com/CosmWasm/cosmwasm-plus/blob/master/contracts/cw20-base/src/enumerable.rs#L45-L51

Note that you can use `Option<String>` for the `start_after` as that is the typical id. Otherwise the same. (I need to write up this pattern in cw0... just testing it out with various contracts before moving it into a standard.

Reply:
I was thinking about this and a good idea would be to encapsulate the pagination functionality inside `range()` directly, or by using an additional function like `interval()`. And with a couple extra options like `Inclusive` (default), `Exclusive`, `OpenBelow`, `OpenAbove`. Limits can be handled the same way: with an option like `Option<usize>`, or with naming (`range()` vs. `range_all()`, etc.)

One of the few things I don't like about Rust is lack of default parameters, by the way (other being the excessive need for `clone()`).

Reply:
Also, I don't like the approach in `calc_range_start()`. Maybe it's better to just get an inclusive range, and then filter out the first / last elements, in case of open intervals / exclusive ranges.

The logic can be a bit complex or cumbersome, but it needs to be written just once, and it can be thoroughly tested. 

It would probably be easier to write a general function using generic types this way, too.